tests/Unit/controller/FrontController/ProductControllerTest.php

```diff
@@ -0,0 +1,310 @@
+                $data['specific_prices'],
+                $data['price'],
+                $data['tax_rate'],
+                $data['ecotax_amount'],
+            ]
+        );
+
+        $priceFormatter = new PriceFormatter();
+
+        foreach ($data['expected'] as $expectedLevel => $expectedValues) {
+            $this->assertArrayHasKey($expectedLevel, $result);
+            foreach ($expectedValues as $expectedKey => $expectedValue) {
+                $this->assertArrayHasKey($expectedKey, $result[$expectedLevel]);
+                $this->assertEquals($priceFormatter->format($expectedValue), $result[$expectedLevel][$expectedKey]);
+            }
+        }
+    }
+
+    public function specificPricesProvider()
+    {
+        return [
+            // EUR to USD, without ecotax
+            [
+                [
+                    'price'           => 31.2,
+                    'tax_rate'        => 20,
+                    'ecotax_amount'   => 0,
+                    'currency'        => [
+                        'conversion_rate' => 1.3,
+                        'sign'            => '$',
+                        'code'            => 'USD',
+                    ],
+                    'specific_prices' => [
+                        0 => [
+                            'id_specific_price'      => '9',
+                            'id_specific_price_rule' => '0',
+                            'id_cart'                => '0',
+                            'id_product'             => '10',
+                            'id_shop'                => '1',
+                            'id_shop_group'          => '0',
+                            'id_currency'            => '0',
+                            'id_country'             => '0',
+                            'id_group'               => '0',
+                            'id_customer'            => '0',
+                            'id_product_attribute'   => '0',
+                            'price'                  => '15.000000',
+                            'from_quantity'          => '15',
+                            'reduction'              => 0,
+                            'reduction_tax'          => '1',
+                            'reduction_type'         => 'amount',
+                            'from'                   => '0000-00-00 00:00:00',
+                            'to'                     => '0000-00-00 00:00:00',
+                            'score'                  => '48',
+                            'quantity'               => '15',
+                            'reduction_with_tax'     => 0,
+                            'nextQuantity'           => -1,
+                        ],
+                    ],
+                    'expected'        => [
+                        [
+                            'discount' => 7.80,
+                            'save'     => 117.00,
+                        ],
+                    ],
+                ],
+            ],
```

Comment:
I see a lot of duplicated data across tests. Maybe these blocks could be generated by a helper method that receives the relevant values only?